### PR TITLE
Implement new activity indicator for iOS14

### PR DESCRIPTION
**Pull request by ctrysbita**
## Description

Apple has updated the look of Activity Indicator in iOS14.

This PR introduced a flag `iOSVersionStyle` for `CupertinoActivityIndicator` to enable new look.
The flag will be disabled by default until the official release of iOS14.

## Related Issues

#60047 

## Tests

I added the following tests:

- Goden file for Activity Indicator with new style.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Piinks** [comment]
cc/ @LongCatIsLooong and @justinmc We recently updated the cupertino activity indicator in https://github.com/flutter/flutter/pull/58392. Is this different from the one that landed two weeks ago?

**Comment by justinmc** [comment]
It looks like this will update us to the iOS 14 appearance, where the previous PR gave us some features that were missing from the iOS 13 version.  No wasted work though, as it appears that the iOS 14 version is very similar to our now fully functional iOS 13 version.  See https://github.com/flutter/flutter/issues/60047.  CC @edwardaux 

@ctrysbita Quickly checking out your branch and using the [Flutter gallery pull-to-refresh demo](https://github.com/flutter/gallery/blob/master/lib/demos/cupertino/cupertino_refresh_demo.dart), I still see the 12 pointed refresh indicator instead of the new 8 pointed one.

Screen Shot 2020-07-06 at 2 22 50 PM

Here's my full code slightly modified from the gallery demo:

```dart
// Copyright 2019 The Flutter team. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:math' show Random;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return CupertinoApp(
      title: 'Cupertino App',
      home: CupertinoRefreshControlDemo(),
    );
  }
}

class CupertinoRefreshControlDemo extends StatefulWidget {
  const CupertinoRefreshControlDemo();

  @override
  _CupertinoRefreshControlDemoState createState() =>
      _CupertinoRefreshControlDemoState();
}

class _CupertinoRefreshControlDemoState
    extends State<CupertinoRefreshControlDemo> {
  static const listCount = 20;
  var randomList = List<int>.generate(listCount, (i) => i + 1);

  void _shuffleList() => randomList.shuffle(Random());

  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      child: CustomScrollView(
        // If left unspecified, the [CustomScrollView] appends an
        // [AlwaysScrollableScrollPhysics]. Behind the scene, the ScrollableState
        // will attach that [AlwaysScrollableScrollPhysics] to the output of
        // [ScrollConfiguration.of] which will be a [ClampingScrollPhysics]
        // on Android.
        // To demonstrate the iOS behavior in this demo and to ensure that the list
        // always scrolls, we specifically use a [BouncingScrollPhysics] combined
        // with a [AlwaysScrollableScrollPhysics]
        physics: const BouncingScrollPhysics(
          parent: AlwaysScrollableScrollPhysics(),
        ),
        slivers: [
          CupertinoSliverNavigationBar(
            automaticallyImplyLeading: false,
            largeTitle: Text(
              'Title',
            ),
          ),
          CupertinoSliverRefreshControl(
            onRefresh: () {
              return Future<void>.delayed(const Duration(seconds: 1))
                ..then<void>((_) {
                  if (mounted) {
                    setState(() => _shuffleList());
                  }
                });
            },
          ),
          SliverList(
            delegate: SliverChildBuilderDelegate(
              (context, index) {
                final title = 'Ok! $index';
                return Card(child: Text(title));
              },
              childCount: listCount,
            ),
          ),
        ],
      ),
    );
  }
}
```

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (edwardaux) [inline packages/flutter/lib/src/cupertino/activity_indicator.dart:79]: I feel like calling it `useNewStyle` isn't very future proof... eg. what will happen when Apple changes it again in iOS 15?

Perhaps a better solution might be to create an enum something like:
```
enum ActivityIndicatorVariant {
  iOS13AndEarlier,
  iOS14,
}
```
I'm not super happy with the naming above, but hopefully the intent is clear.

And whatever naming convention we go with, there's a bunch of other cases where code introduced in this PR talks about the "new" style that should be updated to use the new naming (eg. the names of the tests, the array of alpha values, etc)

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (edwardaux) [inline packages/flutter/lib/src/cupertino/activity_indicator.dart:142]: You've removed `_kTickCount` (which is reasonable given the change), but this comment still refers to it.

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (edwardaux) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart:79]: It would probably also be worth adding a comment to each enum case to describe the specific behaviour

File: packages/flutter/test/cupertino/activity_indicator_test.dart
Comment (edwardaux) [inline packages/flutter/test/cupertino/activity_indicator_test.dart:82]: Nitpick: Not sure what triggered these changes. Have the Flutter formatting rules changed, or is this just based on  the auto-formatting preferences from whatever IDE the PR author is using?  There's a bunch of formatting changes in this PR that mask the actual changes under the covers.

**Comment by LongCatIsLooong** [comment]
@ctrysbita we have a policy that [widget libraries should follow the latest OEM behavior](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#widget-libraries-follow-the-latest-oem-behavior), I think that policy justifies changing the current default appearance of the activity indicator when iOS 14 is out of beta. 
I'm a little bit concerned that in 2 months the new flag introduced in the PR will have to be deprecated. Maybe we should wait until the official release of iOS 14 and then just change the appearance of the activity indicator without the flag?

**Comment by jonahwilliams** [comment]
> Maybe we should wait until the official release of iOS 14 and then just change the appearance of the activity indicator without the flag?

If you do this, most applications will not be updated/installed in time for the release. Then all flutter applications will lag behind the native counterparts

**Comment by justinmc** [comment]
When I was working on iOS 13 features in advance of the release, I just merged them early and no one seemed to have a problem with it.  I vote for just removing the flag if no one sees a specific problem.

File: packages/flutter/test/cupertino/activity_indicator_test.dart
Comment (ctrysbita) [reply packages/flutter/test/cupertino/activity_indicator_test.dart:82]: The file is formated using dartfmt. I think the diff tool should be able to show the actual change in a darker color.

**Comment by ctrysbita** [comment]
> When I was working on iOS 13 features in advance of the release, I just merged them early and no one seemed to have a problem with it. I vote for just removing the flag if no one sees a specific problem.

However, iOS 14 is not officially release yet. Maybe we can use an enum to define the style.

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (ctrysbita) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart]: Fixed in latest push.

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (ctrysbita) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart]: I've added the enum `CupertinoActivityIndicatorStyle` to control the style.
The default style still remain iOS13 until the official release of iOS14.

**Comment by ctrysbita** [comment]
> It looks like this will update us to the iOS 14 appearance, where the previous PR gave us some features that were missing from the iOS 13 version. No wasted work though, as it appears that the iOS 14 version is very similar to our now fully functional iOS 13 version. See #60047. CC @edwardaux
> 
> @ctrysbita Quickly checking out your branch and using the [Flutter gallery pull-to-refresh demo](https://github.com/flutter/gallery/blob/master/lib/demos/cupertino/cupertino_refresh_demo.dart), I still see the 12 pointed refresh indicator instead of the new 8 pointed one. Let me know if I'm doing something wrong or what the plan is for implementing the refresh indicator as seen in #60047.
> 
> Screen Shot 2020-07-06 at 2 22 50 PM
> 
> Here's my full code slightly modified from the gallery demo:

@justinmc The `CupertinoSliverRefreshControl` uses a builder to build refresh indicator. The default builder was set to build the default `CupertinoActivityIndicator`, which is still under iOS13 style.
Currently you have to implement a custom builder if you want to build the activity indicator with iOS14 style inside `CupertinoSliverRefreshControl`.

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/activity_indicator.dart:22]: I recommend that we remove this enum and `style` flags and only support iOS 14.  If we wait a few weeks to merge this PR so that it misses the next stable release, then the following stable release should happen pretty close to September and the likely release of iOS 14.

Flutter's usual policy is to support only the latest version of a design style.  For example, see https://github.com/flutter/flutter/pull/36087 where switches were updated to match iOS 13's style several months before the release of iOS 13.

Otherwise, if we really want to support both versions right now, we need to have a plan for deprecation. That probably involves creating another issue and adding TODOs to the code saying that it needs to be deprecated. Then someone needs to merge a PR to deprecate this, and then later another PR to remove the iOS 13 code altogether.

**Comment by justinmc** [comment]
I was able to see the new indicator by using the `builder` parameter and setting the `style` parameter to iOS 14 👍

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (ctrysbita) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart:22]: I think developers need time to adapt the app for iOS14 and user also need time to update to iOS14. It will be better to support the new style before official release of iOS14 to give developers enough time to adapt their UI.

I've added a TODO for changing default style to iOS14. (Maybe I will do that.)

Comment (justinmc) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart:22]: Sounds good, if you want to do it this way then I'm on board with that.  Thanks for adding the TODO.  I think the deprecation process should look like this:

Roughly 1 month before the release of iOS 14, we should merge another PR that adds the `@deprecated` flag to the `style` parameter.  This PR should also change the default value to `CupertinoActivityIndicatorStyle.iOS14`.  This will have the effect of automatically updating users to iOS 14 style if they haven't specifically set their style.  The PR should be merged 1 month before iOS 14's release so that the PR is released to stable around the same time.

After iOS 14 is released and everything is working well with the new indicator style, we should merge another PR that removes the `style` parameter and `CupertinoActivityIndicatorStyle` altogether.  This will force everyone to the iOS 14 style.

Comment (justinmc) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart:22]: We can either keep https://github.com/flutter/flutter/issues/60047 open until the final step is complete, or we can create 2 new issues to track the two points above.

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/activity_indicator.dart:41]: `style` is ambiguous with other unrelated things in Flutter (like [TextStyle](https://master-api.flutter.dev/flutter/cupertino/CupertinoTextField/style.html)).  I think we should rename `style` and `CupertinoActivityIndicatorStyle`.  Maybe to something like `iOSVersionStyle` and `CupertinoActivityIndicatorIOSVersionStyle`?

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/activity_indicator.dart:23]:
```suggestion
  /// The style that is used in iOS13 and earlier (12 points).
```

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/activity_indicator.dart:26]:
```suggestion
  /// The style that was introduced in iOS14 (8 points).
```

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/activity_indicator.dart:87]:
```suggestion
  /// Defaults to [CupertinoActivityIndicatorStyle.iOS13].
```

**Review by justinmc — COMMENTED** [review]
Here are a few improvements if we're going to take the flag approach here.

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (ctrysbita) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart]: Sure, we can keep the old issue and update the progress.

File: packages/flutter/lib/src/cupertino/activity_indicator.dart
Comment (ctrysbita) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart]: I've renamed it to `iOSVersionStyle`.

**Comment by ctrysbita** [comment]
Changes are rebased on the latest `master` branch.

**Comment by justinmc** [comment]
@ctrysbita Can you paste some code for how to test this widget? I'm having trouble using CupertinoSliverRefreshControl.

Otherwise I think this PR is pretty much good!  Thanks for making those changes.

**Comment by ctrysbita** [comment]
> @ctrysbita Can you paste some code for how to test this widget? I'm having trouble using CupertinoSliverRefreshControl.
> 
> Otherwise I think this PR is pretty much good! Thanks for making those changes.

Actually you can build an activity indicator directly without `CupertinoSliverRefreshControl`.

```dart
import 'package:flutter/cupertino.dart';

main(List<String> args) {
  runApp(CupertinoApp(
    home: CupertinoPageScaffold(
      child: Center(
        child: CupertinoActivityIndicator(
          iOSVersionStyle: CupertinoActivityIndicatorIOSVersionStyle.iOS14,
          radius: 20,
        ),
      ),
    ),
  ));
}
```

**Review by justinmc — APPROVED** [review]
LGTM 👍 

I got it working locally with your code above, thanks for the help.

I did notice some subtle differences between the native activity and this one in my testing on the iOS 14 beta.  The animations seem slightly more complex on native at the moment where the indicator starts spinning.  I think we can improve this as we go, though, and this PR is a great first step.

Thanks for the contribution!  Please follow up with the deprecation of iOSVersionStyle when the time comes!  If you tag me in the PR I'll be happy to review that one too.

**Comment by jmagman** [comment]
Deprecation follow up at https://github.com/flutter/flutter/issues/62521.

**Pull request merged**
